### DISCUSSION TITLE
Case insensitive orderBy

### DISCUSSION THREAD

**Discussion by ericchernuka** [mswjs/data#194] [Ideas]
I'm just running into this and thought I'd see if there's any interest in having the ability to use a `localeCompare` option on the orderBy rather than [`>`|`<`](https://github.com/mswjs/data/blob/d32f15461fc0ef915da6709de62e9e9156b8a234/src/query/sortResults.ts#L79) on the string. Just running into situations where things are out of order due to casing.

![image](https://user-images.githubusercontent.com/243155/155815415-b69c1be7-1ed6-42fa-bde8-ec604767a6df.png)

**Comment by kettanaito** [comment]
Hey, @ericchernuka. Thanks for suggesting this!

I think this is a great idea. Our API is heavily inspired by Prisma, and they have a similar issue open (https://github.com/prisma/prisma/issues/5068). We, however, don't have support for something like a `mode` property on sorting. 

What do you think about the following API?

```js
db.user.findMany({
  where: criteria,
  orderBy: {
    username: {
      direction: 'asc',
      mode: 'insensitive'
    }
  }
})
```

I'm not a fan of complicating the `orderBy` API to be either a `SortDirection` enum or an object of `{ direction: SortDirection, mode?: SortMode }` though. But this may be a sensible way to do it from the user's perspective. 

I'd like to hear your opinion on how you'd expect this API to look.

**Reply by ericchernuka** [reply]
Thanks for the quick reply. I also agree with you about complicating the `orderBy` API, but can't think of another way to achieve this other than making `insensitiveOrderBy` which just adds more surface area to the query API.

I did a quick spin through the Prisma issue and how they're tackling it in filters for Postgres. I think your idea is probably the best solution to be honest.

**Reply by kettanaito** [reply]
We can try implementing it with this ambiguous `orderBy` value as initial concept:

```js
{
  // Preserve the current behavior.
  orderBy: {
    age: 'asc'
  }
}
```

```js
{
  // Add a new one.
  orderBy: {
    name: {
      direction: 'asc',
      mode: 'insensitive'
    }
  }
}
```

Would you be interested in giving me a hand with this?

**Reply by ericchernuka** [reply]
Definitely. Love to help!

**Reply by micha149** [reply]
Hey. I came across a similar Problem in a list ordered by Names where the lowercase Names are not in the alphabetical order where one may expect them. I also thought about a solution, looked how prisma handles this issue, came to the open issue and noticed that their sorting api works different than ours. Further I asked me why this is working in mysql and came to the collation of the tables which are `utf8_unicode_ci` (the `_ci` stands for case insensitive). Then I wondered if its feasable to just introduce new sorting variants and keep the current api as it is.

```ts
{
  orderBy: {
    // Sort with the current case sensitive behavior.
    age: 'asc'
    // Sort with case insensitive behavior.
    name: 'asc_ci'
  }
}
```

**Reply by kettanaito** [reply]
@micha149, that sounds like a practical option to have but at the current stage I think it'd complicate the Data API. I think we should start slow by supporting `orderBy` objects with `direction` and `mode`, and once we learn from that we can incrementally bring new API to handle more complex scenarios. 

I'd also appreciate some help with this `orderBy` change if anybody has time.

**Reply by ericchernuka** [reply]
@kettanaito Apologies. Still planning on getting to this. Work got a little crazy over the past few months so I haven’t had time to start on it.

**Reply by kettanaito** [reply]
No need to apologize! I'm thankful for your interest on this. Let's improve things together!

**Reply by micha149** [reply]
To be honest. My thought was to have less impact to the api, because its just two more strings. I am also fine with a object to describe the orderBy behavior.

**Reply by micha149** [reply]
OK I had a quick look to the source code and wondered what the mode could be besides `default` and `insensitive`. I thought `natural` could also be useful. After having a look in [`String.localeCompare`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare) I noticed the hint to better use [`Intl.Collator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator/Collator) for large arrays. It has [way more options](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator/Collator#options) to consider for sorting like `numeric` or how `a` and `á` has to be compared. Should we take this into account or just implement the `insensitive` mode by calling `.toLowercase()` on the values?

**Reply by kettanaito** [reply]
Yeah, we should begin by just converting the values to the same case. We can always improve on top of that, if necessary.
